evcxr/src/eval_context.rs

```diff
@@ -508,6 +508,31 @@ impl EvalContext {
         Ok(completions)
     }
 
+    pub fn hover(
+        &mut self,
+        code: &str,
+        state: &mut ContextState,
+    ) -> Result<String> {
+        let (modified_code, hover_offset) = if code == "let" {
+            (String::from("let _ = 1;"), 0)
+        } else {
+            (format!("{};", code), code.len() as u32)
+        };
+        let (user_code, code_info) = CodeBlock::from_original_user_code(&modified_code);
+        let user_code = state.apply(user_code, &code_info.nodes)?;
+        let pad_code = state.analysis_code(user_code);
+        self.analyzer.set_source(pad_code.code_string())?;
+        //if I set the hover position to be the start position of input code, then 
+        //`:doc std::fs::File::create` will return the information of `std`, rather than `create`
+        let wrapped_offset = pad_code.user_offset_to_output_offset_for_hover()? as u32 + hover_offset;
```

Comment:
I tried changing this to the following and it worked for me (including in jupyter):

```rust
let wrapped_offset = pad_code.user_offset_to_output_offset(hover_offset)? as u32;
```

I guess if you're passing `hover_offset` as `code.len()` then you'll be getting the hover for the `;`, so you could consider subtracting 1 to get the last character of what the user typed, but even without subtracting 1 it still seemed to work for me.

Reply:
I test with 
```
let wrapped_offset = pad_code.user_offset_to_output_offset(hover_offset)? as u32;
```
It works fine if I don't include `init.evcxr` and `prelude.rs`. But after I include `init.evcxr` and `prelude.rs`, then I have the problem before, it seems `wrapped_offset` didn't get the right position. This is the sorce code when I include `init.evcxr` and `prelude.rs`:
```
#![allow(unused_imports, unused_mut, dead_code)]
#![allow(unused_imports, unused_variables, non_camel_case_types)]

use serde::Serialize;
use mongodb::Client;
use serde::Deserialize;
use chrono::Timelike;
use chrono::Duration;
use bt::prelude::*;
use bt::std_prelude::*;
use pytest::prelude::*;
use stra::prelude::*;
use getdata::prelude::*;

            struct EvcxrUserCodeError {}
            impl<T: std::fmt::Display> From<T> for EvcxrUserCodeError {
                fn from(error: T) -> Self {
                    eprintln!("{}", error);
                    
                    EvcxrUserCodeError {}
                }
            }
        
fn evcxr_variable_store<T: 'static>(_: T) {}
#[allow(unused_variables)]
async fn evcxr_analysis_wrapper(
) -> Result<(), EvcxrUserCodeError> {
i32;
Ok(())
}
```

Reply:
Turned out to be an issue with attributes not being properly committed. I've pushed a [fix](https://github.com/evcxr/evcxr/commit/eca319a1f54f7408c03d11640774416528cc4624) it.

A more detailed explanation in case you're interested:

`CodeBlock::user_offset_to_output_offset` relies on the code that the user is currently executing having kind `CodeKind::OriginalUserCode` and any code that the user previously executed having a different kind. The method `EvalContext::commit_old_user_code` commits already-executed user code by changing its origin from `CodeKind::OriginalUserCode` to `CodeKind::OtherUserCode`, however it missed doing so for attributes (like `#![allow(...)]`), which were a relatively recent addition to evcxr. This meant that `CodeBlock::user_offset_to_output_offset` thought that the attribute was part of what was being executed now, so for example offset 0 would be the `#`.